Stop validating schema for labels in k8s operator

hugoShaka opened the pull request:
Proposal to fix #15204

hugoShaka committed:
Stop validating schema for labels (#15204)

marcoandredinis commented on operator/controllers/resources/role_controller_test.go:
```diff
@@ -73,6 +75,62 @@ func TestRoleCreation(t *testing.T) {
+spec:
+  allow:
+    node_labels:
+      'foo': ['bar']
```
I'm confident it should work, but can we please add the example that triggered the issue as test?
```yaml
'*': ['*']
```

hugoShaka committed:
fixup! Stop validating schema for labels (#15204)

codingllama commented on examples/chart/teleport-cluster/charts/teleport-operator/templates/resources.teleport.dev_roles.yaml:
```diff
@@ -40,9 +40,7 @@ spec:
                 properties:
                   app_labels:
                     additionalProperties:
```
What's the implication of dropping `items` and `type`? Could I do something totally wild like `app_labels: 1` ?

hugoShaka replied:
Yes, app_labels is still an object but we don't validate anything in it, so you could totally do
```
app_labels:
  key: value
  key2: ["val1", "val2"]
  key3: {"totally":"invalid"}
  key4: 1
```

This is an obviously terrible solution, the alternative is to change the `utils.Strings` marshalling behaviour. @r0mant  thinks we should not touch `utils.Strings` but @marcoandredinis  and @tigrato don't agree. I'll switch back the PR to draft until we reach a consensus

marcoandredinis replied:
Please note that Teleport would reject the invalid ones
So, while it would pass the K8S validations, changes wouldn't be persisted and an error would be returned to the user

codingllama replied:
I'd be wary of messing with `utils.Strings`, but no strong opinions on my part.

r0mant replied:
Yeah I think changing utils.Strings marshaling is asking for more trouble than needed at this point.

codingllama commented on operator/controllers/resources/role_controller_test.go:
```diff
@@ -73,6 +75,62 @@ func TestRoleCreation(t *testing.T) {
+	namespace := createNamespaceForTest(t, k8sClient)
+	roleName := "interns"
+
+	roleYAML := yaml.NewYAMLOrJSONDecoder(bytes.NewBuffer([]byte(`
```
Do you mind refactoring into a table test, so we can add more examples? (Like the totally absurd one I mentioned earlier?)

codingllama commented on operator/controllers/resources/role_controller_test.go:
```diff
@@ -73,6 +75,62 @@ func TestRoleCreation(t *testing.T) {
+spec:
+  allow:
+    node_labels:
+      '*': ['*']
```
Is `'*': '*'` also OK? Mind adding a test for it?

Also, since we are here, let's test something more reasonable too, like two regular labels and multiple labels in the array.

codingllama commented on operator/controllers/resources/role_controller_test.go:
```diff
@@ -73,6 +75,62 @@ func TestRoleCreation(t *testing.T) {
+
+	var roleX resourcesv5.TeleportRole
+	err := roleYAML.Decode(&roleX)
+	require.NoError(t, err)
```
Where does the test break before the change? Could we add some messages to the require calls so it's easier to distinguish what is setup and what is the focus of the test?

hugoShaka replied:
I refacto-ed the test, is this clearer now ?

codingllama replied:
Better, thanks!

codingllama commented on operator/crdgen/schemagen.go:
```diff
@@ -167,6 +167,19 @@ func (generator *SchemaGenerator) traverseInner(message *Message) (*Schema, erro
+		// and ask Kubernetes to preserve unknown fields.
+		if field.CustomType() == "Labels" {
+			prop.Type = "object"
+			truthy := true
+			prop.AdditionalProperties = &apiextv1.JSONSchemaPropsOrBool{
+				Schema: &apiextv1.JSONSchemaProps{
+					XPreserveUnknownFields: &truthy,
```
```suggestion
			preserveUnknownFields := true
			prop.AdditionalProperties = &apiextv1.JSONSchemaPropsOrBool{
				Schema: &apiextv1.JSONSchemaProps{
					XPreserveUnknownFields: &preserveUnknownFields,
```

codingllama reviewed (commented):
Apologies for the silly Qs, I'm not familiar with this part of the codebase.

codingllama commented:
PR title change is a suggestion, feel free to revert.

hugoShaka committed:
operator: support receiving invalid TeleportRole CRs

This PR makes the controller-runtime manager decode TeleportRole
resource in an unstructured object. This avoid errors during decoding.
The resource is later decoded as a TeleportRole, with proper error
handling.

hugoShaka committed:
Update operator/crdgen/schemagen.go

Co-authored-by: Alan Parra <alan.parra@goteleport.com>

hugoShaka committed:
Lint

hugoShaka commented:
After more testing I had to change the types used by the `TeleportRole` controller. It now decodes everything into an unstructured object and we can "gracefully" handle when invalid resources are sent to us.

I think the PR is now in good-enough shape for another review.

hugoShaka committed:
fixup! operator: support receiving invalid TeleportRole CRs

marcoandredinis commented on operator/controllers/resources/role_controller_test.go:
```diff
@@ -33,57 +39,189 @@ import (
 	resourcesv5 "github.com/gravitational/teleport/operator/apis/resources/v5"
 )
 
+// When I create a TeleportRole CR in Kubernetes, the corresponding TeleportRole must be created in Teleport
+// When I delete a TeleportRole CR in Kubernetes, the corresponding TeleportRole must be deleted in Teleport
```
```suggestion
// When I delete a TeleportRole CR in Kubernetes, the corresponding TeleportRole must be deleted in Teleport
```

marcoandredinis commented on operator/controllers/resources/role_controller_test.go:
````diff
@@ -33,57 +39,189 @@ import (
 	})
 
-	// The role is deleted in K8S
+	// Cleanup and test, we delete the role in Kubernetes```
````
```suggestion
	// Cleanup and test, we delete the role in Kubernetes
```

marcoandredinis commented on operator/controllers/resources/role_controller_test.go:
```diff
@@ -33,57 +39,189 @@ import (
+			// If failure is expected we should not see the resource in Teleport
+			if tc.shouldFail {
+				// We wait 1 second to ensure reconciliation happened
+				time.Sleep(time.Second)
```
Can we replace this `Sleep` with an `Eventually`?

hugoShaka replied:
Not really, because if I understand correctly `Evevntally` will retry gradually. Here I want to assert this does not work after 1 second. If I use eventually it will fail on the first try, maybe before the operator finishes its reconciliation. I think the way to fix this test would be to get the kubernetes object and validate the error condition is here. I'll fix it today.

marcoandredinis replied:
Oh I see.
So, we should ideally wait for the reconciliation to happen :thinking:
Yeah, validating the error condition of the K8s object seems to be a good approach :+1:

marcoandredinis commented on operator/controllers/resources/role_controller_test.go:
```diff
@@ -33,57 +39,189 @@ import (
-	teleportServer, operatorName := defaultTeleportServiceConfig(t)
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			// Creating the Kubernetes resource. We are using an untyped client to be able to create invalid resources.
```
I think we can mark the inner `t *testing.T` test as Parallel 
Each test has its own role name, so it should be safe

codingllama commented on operator/controllers/resources/role_controller_test.go:
```diff
@@ -33,57 +39,189 @@ import (
+		shouldFail   bool
+		expectedSpec *types.RoleSpecV5
+	}{
+		{
+			"Valid login list",
+			[]byte(`
+allow:
+  logins:
+  - ubuntu
+  - root
+`),
+			false,
+			&types.RoleSpecV5{
+				Allow: types.RoleConditions{
+					Logins: []string{"ubuntu", "root"},
+				},
+			},
+		},
```
nit: Use labeled struct initialization? It has better readability and is easier to extend later.

```suggestion
		{
			name: "Valid login list",
			roleSpecYAML: []byte(`
allow:
  logins:
  - ubuntu
  - root
`),
			shouldFail: false,
			expectedSpec: &types.RoleSpecV5{
				Allow: types.RoleConditions{
					Logins: []string{"ubuntu", "root"},
				},
			},
		},
```

Same for others.

codingllama commented on operator/controllers/resources/role_controller_test.go:
```diff
@@ -33,57 +39,189 @@ import (
+	ctx, tClient, k8sClient, ns := setupKubernetesAndTeleport(t)
+	tests := []struct {
+		name         string
+		roleSpecYAML []byte
```
Use `string` instead of `[]byte`?

codingllama commented on operator/controllers/resources/role_controller_test.go:
```diff
@@ -197,6 +326,20 @@ func TestRoleUpdate(t *testing.T) {
 	})
 }
 
+func setupKubernetesAndTeleport(t *testing.T) (context.Context, auth.ClientI, kclient.Client, *v1.Namespace) {
```
Let's wrap the response in a struct, so each component has a name.

Returning a context also looks a bit like a smell - I'd either keep it entirely within the function or get it as a parameter.

hugoShaka replied:
Done, and extended to user tests as well.

codingllama commented on operator/controllers/resources/utils.go:
```diff
@@ -94,3 +98,29 @@ func getReconciliationCondition(err error) metav1.Condition {
 }
+
+func getValidStructureCondition(err error) metav1.Condition {
+	var condition metav1.Condition
+	if err == nil {
+		condition = metav1.Condition{
+			Type:    ConditionTypeValidStructure,
+			Status:  metav1.ConditionTrue,
+			Reason:  ConditionReasonNoError,
+			Message: "Kubernetes CR was successfully decoded.",
+		}
+	} else {
+		condition = metav1.Condition{
+			Type:    ConditionTypeValidStructure,
+			Status:  metav1.ConditionFalse,
+			Reason:  ConditionReasonFailedToDecode,
+			Message: fmt.Sprintf("Failed to decode Kubernetes CR: %s", err),
+		}
+	}
+	return condition
```
+1

codingllama commented on operator/controllers/resources/utils.go:
```diff
@@ -94,3 +98,29 @@ func getReconciliationCondition(err error) metav1.Condition {
 	return condition
 }
+
+func getValidStructureCondition(err error) metav1.Condition {
```
Maybe: getStructureConditionFromError?  I think this deserves a godoc explaining what is the error we are supposed to pass to it.

codingllama commented on operator/controllers/resources/utils.go:
```diff
@@ -94,3 +98,29 @@ func getReconciliationCondition(err error) metav1.Condition {
+	return condition
+}
+
+func getUnstructuredObjectFromGVK(gvk schema.GroupVersionKind) *unstructured.Unstructured {
```
Move this and getValidStructureCondition to role_controller.go? They don't seem to be used in other files.

codingllama replied:
That sounds OK. Thanks.

codingllama reviewed (commented):
Mostly style nits.

hugoShaka committed:
Apply suggestions from code review

Co-authored-by: Alan Parra <alan.parra@goteleport.com>
Co-authored-by: Marco André Dinis <marco.dinis@goteleport.com>

hugoShaka commented on operator/crdgen/schemagen.go:
```diff
@@ -167,6 +167,19 @@ func (generator *SchemaGenerator) traverseInner(message *Message) (*Schema, erro
+		// Labels are relying on `utils.Strings`, which can either marshall as an array of strings or a single string
+		// This does not pass Schema validation from the apiserver, to workaround we don't specify type for those fields
+		// and ask Kubernetes to preserve unknown fields.
+		if field.CustomType() == "Labels" {
```
The issue only happens when we use custom types that don't marshall consistently. As we're inconsistant in our inconsistencies `User.Traits` are not leveraging the `utils.Strings` type, hence everything is ok.

I added similar YAMLSpec tests for users anyway as this will be handy if we need to reproduce issues.

hugoShaka committed:
Review feedback

codingllama commented on operator/controllers/resources/role_controller.go:
```diff
@@ -74,42 +96,68 @@ func (r *RoleReconciler) Delete(ctx context.Context, obj kclient.Object) error {
+	if err != nil {
+		// We update the status conditions on exit and aggregate the eventual error with the original one.
+		return trace.NewAggregate(
+			trace.WrapWithMessage(
+				err,
+				fmt.Sprintf("failed to convert unstructured Object into resource object: %T", k8sResource)),
```
```suggestion
			trace.Wrap(err, "failed to convert unstructured Object into resource object: %T", k8sResource),
```

codingllama commented on operator/controllers/resources/suite_test.go:
```diff
@@ -200,3 +200,38 @@ func validRandomResourceName(prefix string) string {
+	return testSetup{tClient: tClient, k8sClient: k8sClient, namespace: ns}
+}
+
+func teleportCreateDummyRole(ctx context.Context, t *testing.T, roleName string, tClient auth.ClientI) error {
```
nit: Let's be consistent with the order of context.Context and testing.T between this and setupKubernetesAndTeleport.

codingllama replied:
Either ditch the error return from the func or remove the testing.T parameter? As it stands, this always fails instead of returning an error.

codingllama commented on operator/controllers/resources/role_controller_test.go:
```diff
@@ -29,68 +34,203 @@ import (
-	k8sClient := startKubernetesOperator(t, tClient)
-
-	ns := createNamespaceForTest(t, k8sClient)
+	setup := setupKubernetesAndTeleport(t, ctx)
```
Should we free resources like the auth client and k8s client?

Maybe add a Close method to setup so we can immediately follow this with `defer setup.Close()`?

Same for other usages.

codingllama replied:
I think this one is still an issue.

hugoShaka replied:
kube client does not seem to offer a close/cleanup method (or at least I was not able to find one).

I added teleport client & server cleanup in `setupKubernetesAndTeleport`.

codingllama commented on operator/controllers/resources/user_controller_test.go:
```diff
@@ -31,26 +35,26 @@ import (
 
-	require.NoError(t, teleportServer.Start())
-
-	tClient := clientForTeleport(t, teleportServer, operatorName)
-	k8sClient := startKubernetesOperator(t, tClient)
+var TeleportUserGVK = schema.GroupVersionKind{
```
nit: Use private / non-exported variables?

codingllama commented on operator/controllers/resources/role_controller.go:
```diff
@@ -74,10 +98,33 @@ func (r *RoleReconciler) Delete(ctx context.Context, obj kclient.Object) error {
+
+	// We want to update Status.Conditions at the end of the reconciliation, or if the reconciliation fails
+	defer func() {
+		err := r.Status().Update(ctx, k8sResource)
```
The aggregate does looks a bit off, I think we can just log the Update error if it happens. WDYT?

codingllama commented on operator/controllers/resources/user_controller_test.go:
```diff
@@ -65,35 +69,163 @@ func TestUserCreation(t *testing.T) {
+			// If failure is expected we should not see the resource in Teleport
+			if tc.shouldFail {
+				// We wait 1 second to ensure reconciliation happened
+				time.Sleep(time.Second)
```
As the other thread said, let's try and refactor the Sleeps out. This is flakiness waiting to happen.

hugoShaka committed:
Review feedback pt.2

hugoShaka committed:
Lint: Order imports

codingllama commented on operator/controllers/resources/role_controller.go:
```diff
@@ -74,42 +101,81 @@ func (r *RoleReconciler) Delete(ctx context.Context, obj kclient.Object) error {
+	newStructureCondition := getStructureConditionFromError(err)
+	meta.SetStatusCondition(&k8sResource.Status.Conditions, newStructureCondition)
+	if err != nil {
+		// We update the status conditions on exit
+		statusErr := r.Status().Update(ctx, k8sResource)
+		if statusErr != nil && !errors.Is(statusErr, &json.UnmarshalTypeError{}) {
+			log.Errorf("Failed to report error in status conditions: %s", statusErr)
+		}
```
Suggestion:

Extract:

```go
	silentUpdateStatus := func() {
		if err := r.Status().Update(ctx, k8sResource); err != nil {
			log.Errorf("Failed to report error in status conditions: %s", err)
		}
	}
```

and then changes calls to:

```suggestion
		silentUpdateStatus()
```

Note that I dropped the special handling of err; unless we expect UnmarshalTypeError to be _very_ common then we should be OK logging them.

hugoShaka replied:
Done

codingllama commented on operator/controllers/resources/role_controller.go:
```diff
@@ -74,42 +101,81 @@ func (r *RoleReconciler) Delete(ctx context.Context, obj kclient.Object) error {
-	return err
+
+	// We update the status conditions on exit
+	statusErr := r.Status().Update(ctx, k8sResource)
+	return trace.Wrap(statusErr)
```
Simplify?

```suggestion
	return trace.Wrap(r.Status().Update(ctx, k8sResource))
```

hugoShaka replied:
Done

codingllama reviewed (commented):
We're getting there, nice to see the sleeps gone.

Let's cut down on some of the repetition and address the comment about closing test resources, but then I suppose that's it.

hugoShaka committed:
Review feedback pt.3

hugoShaka committed:
fixup! Review feedback pt.3

hugoShaka committed:
Merge branch 'master' into hugo/15204-crd-strings-validation

hugoShaka committed:
Merge branch 'master' into hugo/15204-crd-strings-validation

hugoShaka committed:
Merge branch 'master' into hugo/15204-crd-strings-validation

hugoShaka committed:
Merge branch 'master' into hugo/15204-crd-strings-validation

hugoShaka committed:
Merge branch 'master' into hugo/15204-crd-strings-validation

hugoShaka committed:
Fix panic during teardown due to uninitialized logger
